crates/astria-sequencer/src/sequencer.rs

```diff
@@ -100,13 +110,35 @@ impl Sequencer {
             config.db_filepath.clone(),
             substore_prefixes
                 .into_iter()
-                .map(std::string::ToString::to_string)
+                .map(ToString::to_string)
                 .collect(),
         )
         .await
         .map_err(anyhow_to_eyre)
         .wrap_err("failed to load storage backing chain state")?;
         let snapshot = storage.latest_snapshot();
+        let upgrades =
+            Upgrades::new(&config.upgrades_filepath).wrap_err("failed constructing upgrades")?;
+
+        ensure_historical_upgrades_applied(&upgrades, &snapshot)
+            .await
+            .wrap_err("historical upgrades not applied")?;
+        if let ShouldShutDown::ShutDownForUpgrade {
+            upgrade_activation_height,
+            block_time,
+            hex_encoded_app_hash,
+        } = should_shut_down(&upgrades, &snapshot)
+            .await
+            .wrap_err("failed to establish if sequencer should shut down for upgrade")?
+        {
+            info!(
+                upgrade_activation_height,
+                latest_app_hash = %hex_encoded_app_hash,
+                latest_block_time = %block_time,
+                "shutting down for upgrade"
+            );
+            return Ok(());
+        }
```

Comment:
how does the node restart after shutting down for an upgrade? the code here would always just shut the node down if started up at that height, so would it be to perform some update then change `should_shut_down` to false, then start the node?

Reply:
Yes, it's for a coordinated upgrade, where the current binary halts, and continues to halt on restart (i.e. getting stuck in a crashloop) until an operator replaces the binary with a new one.  The new binary will need to have hard-coded logic inside `fn should_shut_down()` to return `ContinueRunning` for that specific upgrade.